src/main/thrift/parquet.thrift

```diff
@@ -223,6 +223,17 @@ struct Statistics {
     */
    5: optional binary max_value;
    6: optional binary min_value;
+   /** The number of bytes the row/group or page would take if encoded with plain-encoding */
+   7: optional i64 plain_encoded_bytes;
```

Comment:
Seems that only Plain Encoding will be accouted here? Would non-plain encoding have same statistics here (like non-null-size * type->size()`?

Reply:
IIUC, it means total bytes that as if the data is plain-encoded.

Reply:
Can we only count the variable-length bytes in "Plain" here? Other info can be deduced from Type.

Reply:
I'm open to either approach.  IIUC the suggestion here to change the name to something like:
```
/** Optionally set.  But only  set for byte array columns to help applications determine total unencoded/uncompressed size of the page.
   * This is equivalent to PlainEncoding(values) - (num_values_encoded * 4) (i.e. it doesn't include the size
   * needed to record the lengths of the bytes) nor does it include any size to account for nulls.
   */
encoded_byte_array_data_bytes
```

Reply:
I'd rather do this for all primitive types. Otherwise we may end up with different if/else approaches for different types which complicates the logic.

Reply:
However, I hold a different opinion that null should be considered here. 

For example, if most of values are null, we may under-estimate the size and make the split planning much more imprecise.

Although we have `null_count` somewhere (e.g. Statistics or ColumnIndex) but they are optional and may be missing even if `plain_encoded_bytes` here is provided. If we can enforce `null_count` with `plain_encoded_bytes` to appear together, then this is not an issue.

I am not sure if we can slightly modify the definition of `plain_encoded_bytes` to mean the total bytes as if the data is all non-null and plain-encoded. For numeric types, a null value consumes sizeof(T) bytes. For FLBA type, we can simply assume a null value still consumes its original fixed-length. For BYTE_ARRAY type, a null value consume empty length.

That said, I am still open to follow the proposed approach from @emkornfield because it is simple to understand.

Reply:
I think we end up with if/else statements anyways when we start accounting for logical types so I am warming to the idea of only storing bytes for variable width types.  I originally thought plain encoding was useful because it is already a well understood concept within parquet but it might not add value here.

  I agree nulls can be a significant portion of space but I don't think it should be baked into this field and instead handled as a computation on the histograms below.   My rationale for keeping them separate is the space taken really depends on memory representation which can differ by reader.  The actual nullness some system use 1 bit per null indicator, some use one byte and some use 0 because they use sentinel values in data fields.  Similarly some system reserve space in data representations for null values and some do not.  Whether a system leaves space can also vary depending on at what level a null occurs in a nested structure.

I think we should provide guidance that writers should populate all three of the newly proposed fields so estimates can be accurate in the presence of nulls.

Reply:
I agree on putting non-null data and null data into separate fields. The space for null values can have a significant impact on memory footprint so I want to employ these statistics to derive a good batch size while reading data.

It also makes sense to store un-encoded bytes for only variable-length types (in the parquet specs it solely means BYTE_ARRAY type.) But that is not easy to use in these cases:
- Get the total raw size of the file (a.k.a. that size of all columns).
- Get the total raw size of some selected columns.
- Get the total raw size of selected columns in some row groups.
- ...

We need to look at different levels of metadata or even perform some computation to gather the information required above. So my point is to write the raw size info for every data type (with logical type considered) and store/aggregate them into page and column-chunk levels (or even file level?). That would make life easier as the time spent in the planning stage is critical to some analytics use cases.

Reply:
> We need to look at different levels of metadata or even perform some computation to gather the information required above. So my point is to write the raw size info for every data type (with logical type considered) and store/aggregate them into page and column-chunk levels (or even file level?). That would make life easier as the time spent in the planning stage is critical to some analytics use cases.

@wgtmac  would the following changes suffice to address your concerns:
1.  Change the name of the fields to `logical_stored_value_bytes` and define the byte count for each logical type (for Decimal, I'd propose using the underlying size of what it would take to use plain-encoding, for BYTE_ARRAY in this case, for consistency I think this means for BYTE_ARRAY we should also use the amount of space PLAIN_ENCODING would take).
2. Extract the three fields into a new struct  something like:`SizeEstimationStatistics`.
3. In addition to placing this struct into Statistics (which takes care of column level and page level) stats, also put it onto RowGroup? I'd hesitate to put it at the file level because this seems out of character with other metadata) and summing across row groups should be lightweight compared to the overhead of parsing the FileMetadata anyways?
4. (Optional) If we were really concerned about optimizations we could convert the histogram to cumulative distribution function, which would avoid summing to get leaf-nulls.

Reply:
> for Decimal, I'd propose using the underlying size of what it would take to use plain-encoding

Does this mean the result is different if a certain decimal type uses INT32/INT64/FLBA/BA physical type? Should we use the minimal bytes required for its precision?

> Extract the three fields into a new struct something like:SizeEstimationStatistics.

This sounds good. Otherwise we need to check existence of separate fields associated with this patch.

> In addition to placing this struct into Statistics (which takes care of column level and page level) stats, also put it onto RowGroup? I'd hesitate to put it at the file level because this seems out of character with other metadata) and summing across row groups should be lightweight compared to the overhead of parsing the FileMetadata anyways?

I think putting into Statistics is enough. It is cheap to access and compute sum of different columns in the RowGroup from there. Putting into RowGroup is really tricky as we can only sum all columns. It is not realistic to permute combination for all columns.

> If we were really concerned about optimizations we could convert the histogram to cumulative distribution function, which would avoid summing to get leaf-nulls.

If most cases, the nested depth should not be too deep. I think we can make it simple as the current histogram design. IIUC, we can also derive the cumulative result easily from the current histogram design.

Reply:
> Does this mean the result is different if a certain decimal type uses INT32/INT64/FLBA/BA physical type? Should we use the minimal bytes required for its precision?

Updated to use minimal bytes.  This  actually makes me want to revert back only variable length bytes exclusive of length.  Even though this requires some computation on readers size, I think it is probably the most worthwhile again because of different memory models and schema evolution.  For instance if precision of the decimal type is widened in a data set, I think for join planning and memory allocation purposes it is most useful to use the reader precision rather then the writer precision.  Similarly in common memory models (e.g. spark unsafe row) integers always take 64 bits regardless of the storage type.

> Putting into RowGroup is really tricky as we can only sum all columns. It is not realistic to permute combination for all columns.

Yep, came to the same conclusion as I was updating the PR, this is impossible and would need other solutions.

> If most cases, the nested depth should not be too deep. I think we can make it simple as the current histogram design. IIUC, we can also derive the cumulative result easily from the current histogram design.

reverted back to historgram.